src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/Utf8BufferTextWriter.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    internal sealed class Utf8BufferTextWriter : TextWriter
```

Comment:
You should use the thread static trick that Utf8BufferTextReader uses. You might also want to look at using `_utf8NoBom.GetEncoder()` instead of GetByteCount and GetBytes

Reply:
Hmmm, the writer might need `GetEncoder` to handle surrogate chars. I'll unit test to see what happens.

You want thread static now or later? I thought you didn't like feature creep? 😋

Reply:
> You want thread static now or later? I thought you didn't like feature creep? 😋

I think you'll need it once you start allocating the encoder per Utf8BufferTextWriter. That's why I added it to the reader. It'll probably actually help the performance a bunch (I hope 😄 ).